### PR TITLE
Fix failing translations for hedy exceptions #1804

### LINKED ISSUE #1804: [BUG] ServerError error message instead of "Wrong Level" error message (when in German)
**Describe the bug**
When you use `echo` in level 2 _in German_, you get the `ServerError` error message instead of the `Wrong Level` error message.

**Paste the Hedy code & level**
Just
```
echo
```
or (with same effect) something more potentially meaningful containing lines beginning with `echo` such as
```
antwort is ask Wie heißt du?
echo Aha, du heißt
```
Level **2**, with **`Deusch`** chosen in the language selector on the top right of the website.

**Add a screenshot (optional)**
image

> ### Hoppla! Vielleicht haben wir einen kleinen Fehler gemacht.
>
> Du hast ein Programm geschrieben das wir nicht erwartet haben. Wenn du uns hlefen möchtest, dann sende eine Email an hedy@felienne.com und gib darin das Level an in dem du dieses Programm geschrieben hast und füge den Programmcode hinzu. In der Zwischenzeit, probiere aus es ein bisschen anders zu formulieren und schau nochmal in die Beispiele. Danke!

<details><summary>relevant server terminal output (Tracebacks, click to expand)</summary>

```
Traceback (most recent call last):
  File "/app/hedy.py", line 2207, in transpile_inner
    python = transpiler(punctuation_symbols, lookup_table).transform(abstract_syntax_tree)
  File "/usr/local/lib/python3.9/site-packages/lark/visitors.py", line 130, in transform
    return self._transform_tree(tree)
  File "/usr/local/lib/python3.9/site-packages/lark/visitors.py", line 125, in _transform_tree
    children = list(self._transform_children(tree.children))
  File "/usr/local/lib/python3.9/site-packages/lark/visitors.py", line 116, in _transform_children
    yield self._transform_tree(c)
  File "/usr/local/lib/python3.9/site-packages/lark/visitors.py", line 125, in _transform_tree
    children = list(self._transform_children(tree.children))
  File "/usr/local/lib/python3.9/site-packages/lark/visitors.py", line 116, in _transform_children
    yield self._transform_tree(c)
  File "/usr/local/lib/python3.9/site-packages/lark/visitors.py", line 126, in _transform_tree
    return self._call_userfunc(tree, children)
  File "/usr/local/lib/python3.9/site-packages/lark/visitors.py", line 97, in _call_userfunc
    raise VisitError(tree.data, tree, e)
lark.exceptions.VisitError: Error trying to process rule "error_echo_dep_2":

Wrong Level

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/app.py", line 469, in parse
    transpile_result = transpile_add_stats(code, level, lang)
  File "/app/app.py", line 539, in transpile_add_stats
    result = hedy.transpile(code, level, lang_)
  File "/app/hedy.py", line 1904, in transpile
    transpile_result = transpile_inner(input_string, level, lang)
  File "/app/hedy.py", line 2216, in transpile_inner
    raise E.orig_exc
  File "/usr/local/lib/python3.9/site-packages/lark/visitors.py", line 93, in _call_userfunc
    return f(children)
  File "/app/hedy.py", line 1220, in error_echo_dep_2
    raise hedy.exceptions.WrongLevelException(1,  'echo', "echo_out")
exceptions.WrongLevelException: Wrong Level
INFO:werkzeug:172.17.0.1 - - [26/Jan/2022 00:10:28] "POST /parse HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/app/hedy.py", line 2207, in transpile_inner
    python = transpiler(punctuation_symbols, lookup_table).transform(abstract_syntax_tree)
  File "/usr/local/lib/python3.9/site-packages/lark/visitors.py", line 130, in transform
    return self._transform_tree(tree)
  File "/usr/local/lib/python3.9/site-packages/lark/visitors.py", line 125, in _transform_tree
    children = list(self._transform_children(tree.children))
  File "/usr/local/lib/python3.9/site-packages/lark/visitors.py", line 116, in _transform_children
    yield self._transform_tree(c)
  File "/usr/local/lib/python3.9/site-packages/lark/visitors.py", line 125, in _transform_tree
    children = list(self._transform_children(tree.children))
  File "/usr/local/lib/python3.9/site-packages/lark/visitors.py", line 116, in _transform_children
    yield self._transform_tree(c)
  File "/usr/local/lib/python3.9/site-packages/lark/visitors.py", line 126, in _transform_tree
    return self._call_userfunc(tree, children)
  File "/usr/local/lib/python3.9/site-packages/lark/visitors.py", line 97, in _call_userfunc
    raise VisitError(tree.data, tree, e)
lark.exceptions.VisitError: Error trying to process rule "error_echo_dep_2":

Wrong Level

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/app.py", line 469, in parse
    transpile_result = transpile_add_stats(code, level, lang)
  File "/app/app.py", line 539, in transpile_add_stats
    result = hedy.transpile(code, level, lang_)
  File "/app/hedy.py", line 1904, in transpile
    transpile_result = transpile_inner(input_string, level, lang)
  File "/app/hedy.py", line 2216, in transpile_inner
    raise E.orig_exc
  File "/usr/local/lib/python3.9/site-packages/lark/visitors.py", line 93, in _call_userfunc
    return f(children)
  File "/app/hedy.py", line 1220, in error_echo_dep_2
    raise hedy.exceptions.WrongLevelException(1,  'echo', "echo_out")
exceptions.WrongLevelException: Wrong Level

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2464, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2450, in wsgi_app
    response = self.handle_exception(e)
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1867, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.9/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.9/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/app/app.py", line 508, in parse
    response = hedy_error_to_response(ex, hedy_errors)
  File "/app/app.py", line 565, in hedy_error_to_response
    "Error": translate_error(ex.error_code, translations, ex.arguments),
  File "/app/app.py", line 589, in translate_error
    return error_template.format(**arguments)
KeyError: 'original_level'
```

</details>

tl;dr: Root cause seems to be `KeyError: 'original_level'`

**Expected behavior**
The `Wrong Level` error message should be shown instead of the `ServerError` error message. According to https://www.hedycode.com/translate/en/de , that'd currently be

> Das war richtige Hedy-Sprache, aber nicht im richtigen Level. Du hast ein Sprachelement aus Level 1 in Level 2 verwendet.

**What machine and browser you were using (optional)**
Browser: Firefox 96.0.1 (64-bit)

Output of `nix-shell -p nix-info --run "nix-info -m"`:
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.92, NixOS, 22.05 (Quokka)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.5.1`
 - channels(root): `"nixos-22.05pre347261.6d8215281b2, nixos-unstable-21.03pre257780.e9158eca70a"`
 - channels(das-g): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Reproduced on https://www.hedycode.com/hedy/2 and locally with the Docker instructions on 025068018d0bdfea0ddcd16962e8ef3a4fcba36b.

### PR DESCRIPTION
**Pull request by boryanagoncharenko**
**Fix failing translations for all HedyExceptions**

* _Changes the translations of exceptions which did not have correct keywords and failed with KeyError when translated_
* _Adds automated tests that ensures no subclasses of HedyException fails because of wrong arguments. The tests are parametrized to automatically include newly added languages and exceptions._

**Fixes #1804 **

**How to test**

* To ensure the tests are passing, some of the already translated error messages had to be changed back to English. Is this ok? Maybe there is a nicer way to adapt the translation so that we do not completely lose it?
* Check how the tests are created. The idea of the tests is that they will be automatically created when a new language or exception is created but is the code easy to understand? Is the name of each individual test clear?

### CODE REVIEW

**Comment by Felienne** [comment]
Thanks for looking at this long open issue @boryanagoncharenko!!

> To ensure the tests are passing, some of the already translated error messages had to be changed back to English. Is this ok? Maybe there is a nicer way to adapt the translation so that we do not completely lose it?

Do you happen to have a list of which you reverted (looking at all 29 files involves a lot of clicking and scrolling)

I sometimes use a bit of Google translate to change things for languages I can more or less understand (like everything Germanic is sort of doable) Maybe we can split the languages up and do some manual google translating?

**Comment by boryanagoncharenko** [comment]
I changed my commit to contain only the translation fixes which I am comfortable with (basically Bulgarian) and the fixes which did not lose an existing translation. So the test should fail only for the translations where we could use online services to avoid translation loss. Here is a list of them:

- test_translate_hedy_exception_CyclicVariableDefinitionException_to_ar_lang
- test_translate_hedy_exception_WrongLevelException_to_cs_lang
- test_translate_hedy_exception_WrongLevelException_to_de_lang
- test_translate_hedy_exception_WrongLevelException_to_es_lang
- test_translate_hedy_exception_WrongLevelException_to_fy_lang
- test_translate_hedy_exception_WrongLevelException_to_hu_lang
- test_translate_hedy_exception_WrongLevelException_to_id_lang
- test_translate_hedy_exception_WrongLevelException_to_it_lang
- test_translate_hedy_exception_WrongLevelException_to_pt_BR_lang
- test_translate_hedy_exception_WrongLevelException_to_pt_PT_lang
- test_translate_hedy_exception_WrongLevelException_to_sw_lang

Let me know which languages you are comfortable with and for the rest I will try to use online translation services.

**Comment by boryanagoncharenko** [comment]
It seems that even though this test does fail locally, it always succeeds here. Converting to a draft to investigate the matter.

**Comment by Felienne** [comment]
> It seems that even though this test does fail locally, it always succeeds here. Converting to a draft to investigate the matter.

Curious! What tests fails for you? I thought we had this problem since there was no test covering it. Would be great if you can dive in, thanks!

**Comment by Felienne** [comment]
> > It seems that even though this test does fail locally, it always succeeds here. Converting to a draft to investigate the matter.
> 
> Curious! What tests fails for you? I thought we had this problem since there was no test covering it. Would be great if you can dive in, thanks!

Ah got it, you added a test, but that does not fail on GitHub! Good luck bug hunting ;)

**Comment by boryanagoncharenko** [comment]
The test is now failing as expected! Turned out that we did not compile the babel translations before each unit tests run (which makes sense because we never used them in unit tests before). I added the pybabel compile step to the unit test workflow and it is working as expected.

Now back to the question of which languages we could translate ourselves. Here is the list once again:
-     test_translate_hedy_exception_CyclicVariableDefinitionException_to_ar_lang
-     test_translate_hedy_exception_WrongLevelException_to_cs_lang
-     test_translate_hedy_exception_WrongLevelException_to_de_lang
-     test_translate_hedy_exception_WrongLevelException_to_es_lang
-     test_translate_hedy_exception_WrongLevelException_to_fy_lang
-     test_translate_hedy_exception_WrongLevelException_to_hu_lang
-     test_translate_hedy_exception_WrongLevelException_to_id_lang
-     test_translate_hedy_exception_WrongLevelException_to_it_lang
-     test_translate_hedy_exception_WrongLevelException_to_pt_BR_lang
-     test_translate_hedy_exception_WrongLevelException_to_pt_PT_lang
-     test_translate_hedy_exception_WrongLevelException_to_sw_lang

Are there any languages in the list above that you are comfortable adapting?

**Comment by Felienne** [comment]
> The test is now failing as expected! Turned out that we did not compile the babel translations before each unit tests run (which makes sense because we never used them in unit tests before). I added the pybabel compile step to the unit test workflow and it is working as expected.

Ah great!

> Now back to the question of which languages we could translate ourselves. Here is the list once again:
> 
> * test_translate_hedy_exception_CyclicVariableDefinitionException_to_ar_lang
> * test_translate_hedy_exception_WrongLevelException_to_cs_lang
> * test_translate_hedy_exception_WrongLevelException_to_de_lang
> * test_translate_hedy_exception_WrongLevelException_to_es_lang
> * test_translate_hedy_exception_WrongLevelException_to_fy_lang
> * test_translate_hedy_exception_WrongLevelException_to_hu_lang
> * test_translate_hedy_exception_WrongLevelException_to_id_lang
> * test_translate_hedy_exception_WrongLevelException_to_it_lang
> * test_translate_hedy_exception_WrongLevelException_to_pt_BR_lang
> * test_translate_hedy_exception_WrongLevelException_to_pt_PT_lang
> * test_translate_hedy_exception_WrongLevelException_to_sw_lang
> 
> Are there any languages in the list above that you are comfortable adapting?

I will try my best on most of them, it seems it is mostly smallish changes like the one for Arabic that I just pushed. Maybe Hungarian is doable for you but I don't know how similar it is to Bulgarian.

**Comment by Felienne** [comment]
Ok @boryanagoncharenko as a first order of business I am merging in the newest translations from Weblate to prevent further conflicts. Stand by :)

**Comment by Felienne** [comment]
Secondly, it seems that the fix in https://github.com/Felienne/hedy/pull/3114/commits/4f7623982112969797356ba793f9caebf8857c0d did *not* make the test pass either locally or here, but I will deal with that in a bit.

**Comment by Felienne** [comment]
> Ok @boryanagoncharenko as a first order of business I am merging in the newest translations from Weblate to prevent further conflicts. Stand by :)

Merged in! Continuing with the error messages puzzle now :)

**Comment by Felienne** [comment]
ok, I got it, fixed translations coming up!

**Comment by Felienne** [comment]
Ok I did them all apart from Hungarian and Bulgarian, hope you can do both!

**Comment by boryanagoncharenko** [comment]
I think the only remotely close language to Hungarian is Finnish, so I used translate services and I hope the end result isn't something very funny :)

However, it seems that we have a flaky end to end test. This is the same test that failed in the https://github.com/Felienne/hedy/pull/3116 PR.

**Comment by boryanagoncharenko** [comment]
I figured out the issue with the e2e test. When the test gets to update a profile with a wrong birth date, it fails because the text in the translations does not match the supplied key. The text of 'year_invalid' expected a key named {year} but {current_year} was passed instead. It was set correctly only for Dutch and not for the rest of the languages.

It seems that the test started failing when the babel translations were compiled (which happened independently in this PR and also in https://github.com/Felienne/hedy/pull/3116). I will need more time to dive into the e2e tests to figure out why this happens.

**Comment by Felienne** [comment]
> It seems that the test started failing when the babel translations were compiled (which happened independently in this PR and also in #3116). I will need more time to dive into the e2e tests to figure out why this happens.

Pinging @TiBiBa here, maybe good for you to know this and maybe look into it?

**Review by Felienne — APPROVED** [review]
